File: custom_nodes/polygon/polygon_custom_bars_node.py
from typing import Dict, Any, List
import httpx
import logging
from datetime import datetime, timedelta
from nodes.base.base_node import BaseNode
from core.types_registry import get_type, AssetSymbol, OHLCVBar

logger = logging.getLogger(__name__)


class PolygonCustomBarsNode(BaseNode):
    ui_module = "PolygonCustomBarsNodeUI"
    """
    Fetches custom aggregate bars (OHLCV) for a symbol from Polygon.io
    """
    inputs = {"symbol": get_type("AssetSymbol"), "api_key": get_type("APIKey")}
    outputs = {"ohlcv": get_type("OHLCV")}
    default_params = {
        "multiplier": 1,
        "timespan": "day",
        "lookback_period": "3 months",
        "adjusted": True,
        "sort": "asc",
        "limit": 5000,
    }
    params_meta = [
        {"name": "multiplier", "type": "number", "default": 1, "min": 1, "step": 1},
        {"name": "timespan", "type": "combo", "default": "day", "options": ["minute", "hour", "day", "week", "month", "quarter", "year"]},
        {"name": "lookback_period", "type": "combo", "default": "3 months", "options": ["1 day", "3 days", "1 week", "2 weeks", "1 month", "2 months", "3 months", "4 months", "6 months", "9 months", "1 year", "18 months", "2 years", "3 years", "5 years", "10 years"]},
        {"name": "adjusted", "type": "combo", "default": True, "options": [True, False]},
        {"name": "sort", "type": "combo", "default": "asc", "options": ["asc", "desc"]},
        {"name": "limit", "type": "number", "default": 5000, "min": 1, "max": 50000, "step": 1},
    ]

    def _calculate_date_range(self, lookback_period: str) -> tuple[str, str]:
        """Calculate from_date and to_date based on lookback period."""
        now = datetime.now()

        # Parse lookback period
        parts = lookback_period.split()
        if len(parts) != 2:
            raise ValueError(f"Invalid lookback period format: {lookback_period}")

        amount = int(parts[0])
        unit = parts[1].lower()

        # Calculate from_date
        if unit == "day" or unit == "days":
            from_date = now - timedelta(days=amount)
        elif unit == "week" or unit == "weeks":
            from_date = now - timedelta(weeks=amount)
        elif unit == "month" or unit == "months":
            # Approximate months as 30 days
            from_date = now - timedelta(days=amount * 30)
        elif unit == "year" or unit == "years":
            # Approximate years as 365 days
            from_date = now - timedelta(days=amount * 365)
        else:
            raise ValueError(f"Unsupported time unit: {unit}")

        # Format dates as YYYY-MM-DD
        to_date = now.strftime("%Y-%m-%d")
        from_date_str = from_date.strftime("%Y-%m-%d")

        return from_date_str, to_date

    async def execute(self, inputs: Dict[str, Any]) -> Dict[str, List[OHLCVBar]]:
        symbol: AssetSymbol = inputs.get("symbol")
        if not symbol:
            raise ValueError("Symbol input is required")

        api_key = inputs.get("api_key")
        if not api_key:
            raise ValueError("Polygon API key input is required")

        multiplier = self.params.get("multiplier", 1)
        timespan = self.params.get("timespan", "day")
        lookback_period = self.params.get("lookback_period", "3 months")
        adjusted = self.params.get("adjusted", True)
        sort = self.params.get("sort", "asc")
        limit = self.params.get("limit", 5000)

        # Calculate date range based on lookback period
        from_date, to_date = self._calculate_date_range(lookback_period)

        # Construct API URL
        ticker = str(symbol)
        url = f"https://api.polygon.io/v2/aggs/ticker/{ticker}/range/{multiplier}/{timespan}/{from_date}/{to_date}"
        params = {
            "adjusted": str(adjusted).lower(),
            "sort": sort,
            "limit": limit,
            "apiKey": api_key,  # Note: Polygon uses apiKey as query param
        }

        async with httpx.AsyncClient() as client:
            response = await client.get(url, params=params)
            if response.status_code != 200:
                logger.error(f"Polygon API error: {response.status_code} - {response.text}")
                raise ValueError(f"Failed to fetch bars: HTTP {response.status_code}")

            data = response.json()

            if data.get("status") not in ["OK", "DELAYED"]:
                error_msg = data.get("error", "Unknown error")
                raise ValueError(f"Polygon API error: {error_msg}")

            results = data.get("results", [])
            if not results:
                # Return empty list
                return {"ohlcv": []}

            # Convert results to list of dictionaries
            bars = []
            for result in results:
                bar = {
                    "timestamp": result["t"],  # Unix timestamp in milliseconds
                    "open": result["o"],
                    "high": result["h"],
                    "low": result["l"],
                    "close": result["c"],
                    "volume": result["v"],
                }
                # Add optional fields if present
                if "vw" in result:
                    bar["vw"] = result["vw"]  # Volume weighted average price
                if "n" in result:
                    bar["n"] = result["n"]  # Number of transactions
                if "otc" in result:
                    bar["otc"] = result["otc"]  # OTC ticker flag

                bars.append(bar)

            return {"ohlcv": bars}

File: nodes/polygon-custom-bars-node.ts
// src/nodes/custom/polygon/polygon-custom-bars-node.ts

import { Node, NodeCategory, ProgressState, port, type NodeDefinition } from '@sosa/core';
import { AssetClass, AssetSymbol, type OHLCVBar } from './types';

// ======================== Inlined from polygon-service.ts ========================

interface FetchBarsParams {
  multiplier: number;
  timespan: 'minute' | 'hour' | 'day' | 'week' | 'month' | 'quarter' | 'year';
  lookback_period: string;
  adjusted?: boolean;
  sort?: 'asc' | 'desc';
  limit?: number;
}

interface BarMetadata {
  request_id?: string;
  results_count?: number;
  data_status?: string;
  api_status?: string;
  market_open?: boolean;
}

function parseLookbackPeriod(period: string): number {
  const match = period.match(/(\d+)\s*(day|days|week|weeks|month|months|year|years)/i);
  if (!match) {
    return 14;
  }
  const value = parseInt(match[1] ?? '14', 10);
  const unit = (match[2] ?? 'days').toLowerCase();
  switch (unit) {
    case 'day': case 'days': return value;
    case 'week': case 'weeks': return value * 7;
    case 'month': case 'months': return value * 30;
    case 'year': case 'years': return value * 365;
    default: return value;
  }
}

function calculateDateRange(lookbackPeriod: string): { from: number; to: number } {
  const days = parseLookbackPeriod(lookbackPeriod);
  const endDate = new Date();
  const startDate = new Date(endDate.getTime() - days * 24 * 60 * 60 * 1000);
  return { from: startDate.getTime(), to: endDate.getTime() };
}

function isUSMarketOpen(): boolean {
  const now = new Date();
  const etFormatter = new Intl.DateTimeFormat('en-US', {
    timeZone: 'America/New_York',
    hour: 'numeric',
    minute: 'numeric',
    hour12: false,
    weekday: 'short',
  });
  const parts = etFormatter.formatToParts(now);
  const weekday = parts.find((p) => p.type === 'weekday')?.value || '';
  const hour = parseInt(parts.find((p) => p.type === 'hour')?.value || '0', 10);
  const minute = parseInt(parts.find((p) => p.type === 'minute')?.value || '0', 10);
  if (['Sat', 'Sun'].includes(weekday)) {
    return false;
  }
  const currentMinutes = hour * 60 + minute;
  const marketOpenMinutes = 9 * 60 + 30;
  const marketCloseMinutes = 16 * 60;
  return currentMinutes >= marketOpenMinutes && currentMinutes <= marketCloseMinutes;
}

async function fetchBars(
  symbol: AssetSymbol,
  apiKey: string,
  params: FetchBarsParams | Record<string, unknown>
): Promise<[OHLCVBar[], BarMetadata]> {
  const multiplier = (params.multiplier as number) || 1;
  const timespan = (params.timespan as string) || 'day';
  const lookbackPeriod = (params.lookback_period as string) || '3 months';
  const adjusted = params.adjusted !== false;
  const sort = (params.sort as string) || 'asc';
  const limit = (params.limit as number) || 5000;

  const { from, to } = calculateDateRange(lookbackPeriod);

  let ticker = symbol.ticker.toUpperCase();
  if (symbol.assetClass === AssetClass.CRYPTO) {
    ticker = `X:${symbol.toString()}`;
  }

  const url = new URL(
    `https://api.massive.com/v2/aggs/ticker/${ticker}/range/${multiplier}/${timespan}/${from}/${to}`
  );
  url.searchParams.set('apiKey', apiKey);
  url.searchParams.set('adjusted', adjusted.toString());
  url.searchParams.set('sort', sort);
  url.searchParams.set('limit', limit.toString());

  const marketIsOpen = symbol.assetClass === AssetClass.CRYPTO ? true : isUSMarketOpen();

  try {
    const response = await fetch(url.toString());
    if (!response.ok) {
      console.error(`Polygon API error: ${response.status} ${response.statusText}`);
      return [[], { request_id: undefined, results_count: 0, data_status: 'error' }];
    }
    const data = await response.json();
    const apiStatus = data.status;
    if (!['OK', 'DELAYED'].includes(apiStatus) || !data.results) {
      console.warn(`Polygon API returned no results for ${ticker}`);
      return [[], { request_id: data.request_id, results_count: 0, data_status: 'error' }];
    }
    let dataStatus: string;
    if (!marketIsOpen) { dataStatus = 'market-closed'; }
    else if (apiStatus === 'OK') { dataStatus = 'real-time'; }
    else if (apiStatus === 'DELAYED') { dataStatus = 'delayed'; }
    else { dataStatus = 'unknown'; }

    const bars: OHLCVBar[] = data.results.map((r: Record<string, number>) => ({
      timestamp: r.t, open: r.o, high: r.h, low: r.l, close: r.c, volume: r.v,
    }));

    return [bars, {
      request_id: data.request_id,
      results_count: data.resultsCount ?? bars.length,
      data_status: dataStatus,
      api_status: apiStatus,
      market_open: marketIsOpen,
    }];
  } catch (error) {
    console.error(`Error fetching bars for ${ticker}:`, error);
    return [[], { request_id: undefined, results_count: 0, data_status: 'error' }];
  }
}

// ======================== End inlined code ========================

/**
 * Fetches custom aggregate bars (OHLCV) for a symbol from Massive.com API (formerly Polygon.io).
 *
 * Note: Polygon.io has rebranded to Massive.com. The API endpoints have been updated
 * to use api.massive.com, but the API routes remain unchanged.
 *
 * For crypto symbols, the ticker is automatically prefixed with "X:" (e.g., BTCUSD -> X:BTCUSD)
 * as required by the Massive.com crypto aggregates API.
 */
export class PolygonCustomBars extends Node {
  static definition: NodeDefinition = {
    inputs: [port('symbol', 'AssetSymbol')],
    outputs: [port('ohlcv', 'OHLCVBundle')],
    ui: {},
    category: NodeCategory.MARKET,
    requiredCredentials: ['POLYGON_API_KEY'],

    params: [
      { name: 'multiplier', type: 'number', default: 1, min: 1, step: 1 },
      {
        name: 'timespan',
        type: 'combo',
        default: 'day',
        options: ['minute', 'hour', 'day', 'week', 'month', 'quarter', 'year'],
      },
      {
        name: 'lookback_period',
        type: 'combo',
        default: '3 months',
        options: [
          '1 day',
          '3 days',
          '1 week',
          '2 weeks',
          '1 month',
          '2 months',
          '3 months',
          '4 months',
          '6 months',
          '9 months',
          '1 year',
          '18 months',
          '2 years',
          '3 years',
          '5 years',
          '10 years',
        ],
      },
      { name: 'adjusted', type: 'combo', default: true, options: [true, false] },
      { name: 'sort', type: 'combo', default: 'asc', options: ['asc', 'desc'] },
      { name: 'limit', type: 'number', default: 5000, min: 1, max: 50000, step: 1 },
    ],
  };

  protected async run(
    inputs: Record<string, unknown>
  ): Promise<Record<string, unknown>> {
    const symbol = inputs.symbol as AssetSymbol | undefined;

    if (!symbol || !(symbol instanceof AssetSymbol)) {
      throw new Error('Symbol input is required');
    }

    const apiKey = this.credentials.get('POLYGON_API_KEY');
    if (!apiKey) {
      throw new Error('Polygon API key not found in vault');
    }

    const [bars, metadata] = await fetchBars(symbol, apiKey, this.params);

    // Report status via progress callback
    this.emitProgress(
      ProgressState.UPDATE,
      undefined,
      `Fetched bars for ${symbol.ticker}`,
      { polygon_data_status: metadata.data_status || 'unknown' }
    );

    // Build OHLCVBundle (Map with symbol.key as key)
    const bundle = new Map<string, OHLCVBar[]>();
    bundle.set(symbol.key, bars);

    return { ohlcv: bundle };
  }
}
